File: backend/src/websocket/handler.js
import { WebSocketServer } from 'ws';
import {
  getRoomByCode,
  updateRoomHeartbeat,
  leaveRoom,
  closeRoom,
  registerRoomSocket,
  unregisterRoomSocket,
  getRoomSockets,
  getRoomMembers,
  updateRoomPlaybackState
} from './modules/room.js';
import {
  addToQueue,
  getQueue,
  removeFromQueue,
  searchTracks,
  transferPlayback,
  play,
  pause,
  skipToNext,
  skipToPrevious,
  seek,
  getCurrentPlayback
} from './modules/playback.js';
import { getValidAccessToken } from './modules/auth.js';

const clients = new Map(); // socketId -> { ws, userId, roomCode, isHost }

export function initWebSocket(server) {
  const wss = new WebSocketServer({ server, path: '/ws' });

  wss.on('connection', (ws) => {
    const socketId = generateSocketId();
    console.log(`WebSocket connected: ${socketId}`);

    ws.on('message', async (data) => {
      try {
        const message = JSON.parse(data.toString());
        await handleMessage(socketId, ws, message);
      } catch (error) {
        console.error('WebSocket message error:', error);
        sendError(ws, 'Invalid message format');
      }
    });

    ws.on('close', () => {
      handleDisconnect(socketId);
    });

    ws.on('error', (error) => {
      console.error('WebSocket error:', error);
    });
  });

  // Heartbeat checker
  setInterval(() => {
    wss.clients.forEach((ws) => {
      if (ws.isAlive === false) {
        return ws.terminate();
      }
      ws.isAlive = false;
      ws.ping();
    });
  }, 30000);

  wss.on('connection', (ws) => {
    ws.isAlive = true;
    ws.on('pong', () => {
      ws.isAlive = true;
    });
  });

  console.log('WebSocket server initialized');
  return wss;
}

async function handleMessage(socketId, ws, message) {
  const { type, payload } = message;

  switch (type) {
    case 'join_room':
      await handleJoinRoom(socketId, ws, payload);
      break;

    case 'leave_room':
      await handleLeaveRoom(socketId);
      break;

    case 'heartbeat':
      await handleHeartbeat(socketId, payload);
      break;

    case 'search_tracks':
      await handleSearchTracks(socketId, ws, payload);
      break;

    case 'add_to_queue':
      await handleAddToQueue(socketId, payload);
      break;

    case 'remove_from_queue':
      await handleRemoveFromQueue(socketId, payload);
      break;

    case 'playback_control':
      await handlePlaybackControl(socketId, payload);
      break;

    case 'sync_playback':
      await handleSyncPlayback(socketId, payload);
      break;

    case 'transfer_device':
      await handleTransferDevice(socketId, payload);
      break;

    case 'request_token':
      await handleRequestToken(socketId, ws, payload);
      break;

    default:
      sendError(ws, `Unknown message type: ${type}`);
  }
}

async function handleJoinRoom(socketId, ws, payload) {
  const { roomCode, userId, displayName, isHost } = payload;

  try {
    const room = await getRoomByCode(roomCode);
    
    if (!room) {
      return sendError(ws, 'Room not found');
    }

    if (!room.is_active) {
      return sendError(ws, 'Room is not active');
    }

    // Register client
    clients.set(socketId, {
      ws,
      userId,
      roomCode,
      isHost: isHost || false
    });

    // Register in room tracking
    registerRoomSocket(roomCode, socketId, isHost || false);

    // Get room members
    const members = await getRoomMembers(roomCode);
    const queue = await getQueue(roomCode);

    // Send join success
    send(ws, 'room_joined', {
      roomCode,
      room,
      members,
      queue
    });

    // Broadcast to room
    broadcastToRoom(roomCode, 'member_joined', {
      userId,
      displayName,
      isHost: isHost || false
    }, socketId);

    console.log(`${userId} joined room ${roomCode} as ${isHost ? 'host' : 'member'}`);
  } catch (error) {
    console.error('Error joining room:', error);
    sendError(ws, error.message);
  }
}

async function handleLeaveRoom(socketId) {
  const client = clients.get(socketId);
  
  if (!client) {
    return;
  }

  const { roomCode, userId, isHost } = client;

  try {
    await leaveRoom(roomCode, userId);
    
    const role = unregisterRoomSocket(roomCode, socketId);
    
    if (role === 'host') {
      // Host left, close room
      await closeRoom(roomCode);
      broadcastToRoom(roomCode, 'room_closed', {
        reason: 'Host disconnected'
      });
      
      // Disconnect all clients in room
      clients.forEach((c, sid) => {
        if (c.roomCode === roomCode) {
          c.ws.close();
          clients.delete(sid);
        }
      });
    } else {
      // Regular member left
      broadcastToRoom(roomCode, 'member_left', { userId });
    }

    clients.delete(socketId);
    console.log(`${userId} left room ${roomCode}`);
  } catch (error) {
    console.error('Error leaving room:', error);
  }
}

async function handleHeartbeat(socketId, payload) {
  const client = clients.get(socketId);
  
  if (!client || !client.isHost) {
    return;
  }

  const { roomCode } = client;
  await updateRoomHeartbeat(roomCode);
}

async function handleSearchTracks(socketId, ws, payload) {
  const client = clients.get(socketId);
  
  if (!client) {
    return sendError(ws, 'Not in a room');
  }

  const { query } = payload;
  const { userId } = client;

  try {
    const accessToken = await getValidAccessToken(userId);
    const results = await searchTracks(accessToken, query);
    
    send(ws, 'search_results', { results });
  } catch (error) {
    console.error('Error searching tracks:', error);
    sendError(ws, 'Failed to search tracks');
  }
}

async function handleAddToQueue(socketId, payload) {
  const client = clients.get(socketId);
  
  if (!client) {
    return;
  }

  const { roomCode, userId } = client;
  const { track } = payload;

  try {
    const queue = await addToQueue(roomCode, track, userId);
    
    broadcastToRoom(roomCode, 'queue_updated', { queue });
  } catch (error) {
    console.error('Error adding to queue:', error);
  }
}

async function handleRemoveFromQueue(socketId, payload) {
  const client = clients.get(socketId);
  
  if (!client) {
    return;
  }

  const { roomCode, isHost } = client;
  const { queueItemId } = payload;

  if (!isHost) {
    return sendError(client.ws, 'Only host can remove from queue');
  }

  try {
    const queue = await removeFromQueue(roomCode, queueItemId);
    
    broadcastToRoom(roomCode, 'queue_updated', { queue });
  } catch (error) {
    console.error('Error removing from queue:', error);
  }
}

async function handlePlaybackControl(socketId, payload) {
  const client = clients.get(socketId);
  
  if (!client || !client.isHost) {
    return sendError(client?.ws, 'Only host can control playback');
  }

  const { userId, roomCode } = client;
  const { action, deviceId, trackUri, positionMs } = payload;

  try {
    const accessToken = await getValidAccessToken(userId);

    switch (action) {
      case 'play':
        await play(accessToken, deviceId, trackUri, positionMs);
        break;
      case 'pause':
        await pause(accessToken, deviceId);
        break;
      case 'next':
        await skipToNext(accessToken, deviceId);
        break;
      case 'previous':
        await skipToPrevious(accessToken, deviceId);
        break;
      case 'seek':
        await seek(accessToken, positionMs, deviceId);
        break;
    }

    // Broadcast state change
    broadcastToRoom(roomCode, 'playback_changed', {
      action,
      deviceId,
      trackUri,
      positionMs
    });
  } catch (error) {
    console.error('Error controlling playback:', error);
    sendError(client.ws, 'Failed to control playback');
  }
}

async function handleSyncPlayback(socketId, payload) {
  const client = clients.get(socketId);
  
  if (!client || !client.isHost) {
    return;
  }

  const { roomCode } = client;
  const { state } = payload;

  try {
    await updateRoomPlaybackState(roomCode, state);
    
    // Broadcast to all members
    broadcastToRoom(roomCode, 'playback_state', state, socketId);
  } catch (error) {
    console.error('Error syncing playback:', error);
  }
}

async function handleTransferDevice(socketId, payload) {
  const client = clients.get(socketId);
  
  if (!client || !client.isHost) {
    return sendError(client?.ws, 'Only host can transfer device');
  }

  const { userId, roomCode } = client;
  const { deviceId } = payload;

  try {
    const accessToken = await getValidAccessToken(userId);
    await transferPlayback(accessToken, deviceId);
    
    send(client.ws, 'device_transferred', { deviceId });
    
    // Update room state
    await updateRoomPlaybackState(roomCode, { deviceId });
  } catch (error) {
    console.error('Error transferring device:', error);
    sendError(client.ws, 'Failed to transfer device');
  }
}

async function handleRequestToken(socketId, ws, payload) {
  const client = clients.get(socketId);
  
  if (!client) {
    return sendError(ws, 'Not connected');
  }

  const { userId } = client;

  try {
    const accessToken = await getValidAccessToken(userId);
    send(ws, 'token_response', { accessToken });
  } catch (error) {
    console.error('Error getting token:', error);
    sendError(ws, 'Failed to get access token');
  }
}

function handleDisconnect(socketId) {
  const client = clients.get(socketId);
  
  if (client) {
    handleLeaveRoom(socketId);
  }

  clients.delete(socketId);
  console.log(`WebSocket disconnected: ${socketId}`);
}

function send(ws, type, payload) {
  if (ws.readyState === 1) { // OPEN
    ws.send(JSON.stringify({ type, payload }));
  }
}

function sendError(ws, message) {
  send(ws, 'error', { message });
}

function broadcastToRoom(roomCode, type, payload, excludeSocketId = null) {
  clients.forEach((client, socketId) => {
    if (client.roomCode === roomCode && socketId !== excludeSocketId) {
      send(client.ws, type, payload);
    }
  });
}

function generateSocketId() {
  return `ws_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
}

File: backend/websocket/_handler.py

import json
import random
import time

from fastapi import WebSocket, WebSocketDisconnect

from modules.room import (
    get_room_by_code,
    update_room_heartbeat,
    leave_room,
    close_room,
    register_room_socket,
    unregister_room_socket,
    get_room_sockets,
    get_room_members,
    update_room_playback_state,
)
from modules.playback import (
    add_to_queue,
    get_queue,
    remove_from_queue,
    search_tracks,
    transfer_playback,
    play,
    pause,
    skip_to_next,
    skip_to_previous,
    seek,
    get_current_playback,
)
from modules.auth import get_valid_access_token

clients = {}  # socket_id -> { ws, user_id, room_code, is_host }


async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    socket_id = generate_socket_id()
    print(f"WebSocket connected: {socket_id}")

    try:
        while True:
            data = await websocket.receive_text()
            try:
                message = json.loads(data)
            except Exception:
                await send_error(websocket, 'Invalid message format')
                continue

            await handle_message(socket_id, websocket, message)
    except WebSocketDisconnect:
        await handle_disconnect(socket_id)
    except Exception as exc:
        print('WebSocket error:', exc)
        await handle_disconnect(socket_id)


async def handle_message(socket_id, ws, message):
    msg_type = message.get('type')
    payload = message.get('payload') or {}

    if msg_type == 'join_room':
        await handle_join_room(socket_id, ws, payload)
    elif msg_type == 'leave_room':
        await handle_leave_room(socket_id)
    elif msg_type == 'heartbeat':
        await handle_heartbeat(socket_id, payload)
    elif msg_type == 'search_tracks':
        await handle_search_tracks(socket_id, ws, payload)
    elif msg_type == 'add_to_queue':
        await handle_add_to_queue(socket_id, payload)
    elif msg_type == 'remove_from_queue':
        await handle_remove_from_queue(socket_id, payload)
    elif msg_type == 'playback_control':
        await handle_playback_control(socket_id, payload)
    elif msg_type == 'sync_playback':
        await handle_sync_playback(socket_id, payload)
    elif msg_type == 'transfer_device':
        await handle_transfer_device(socket_id, payload)
    elif msg_type == 'request_token':
        await handle_request_token(socket_id, ws, payload)
    else:
        await send_error(ws, f"Unknown message type: {msg_type}")

async def handle_join_room(socket_id, ws, payload):
    room_code = payload.get('roomCode')
    user_id = payload.get('userId')
    display_name = payload.get('displayName')
    is_host = payload.get('isHost') or False

    try:
        room = await get_room_by_code(room_code)
        if not room:
            return await send_error(ws, 'Room not found')
        if not room['is_active']:
            return await send_error(ws, 'Room is not active')

        clients[socket_id] = {
            'ws': ws,
            'userId': user_id,
            'roomCode': room_code,
            'isHost': is_host,
        }

        register_room_socket(room_code, socket_id, is_host)

        members = await get_room_members(room_code)
        queue = await get_queue(room_code)

        await send(ws, 'room_joined', {
            'roomCode': room_code,
            'room': room,
            'members': members,
            'queue': queue,
        })

        await broadcast_to_room(
            room_code,
            'member_joined',
            {'userId': user_id, 'displayName': display_name, 'isHost': is_host},
            exclude_socket_id=socket_id,
        )

        print(f"{user_id} joined room {room_code} as {'host' if is_host else 'member'}")
    except Exception as exc:
        print('Error joining room:', exc)
        await send_error(ws, str(exc))


async def handle_leave_room(socket_id):
    client = clients.get(socket_id)
    if not client:
        return

    room_code = client['roomCode']
    user_id = client['userId']

    try:
        await leave_room(room_code, user_id)
        role = unregister_room_socket(room_code, socket_id)

        if role == 'host':
            await close_room(room_code)
            await broadcast_to_room(room_code, 'room_closed', {'reason': 'Host disconnected'})

            to_close = [sid for sid, c in clients.items() if c['roomCode'] == room_code]
            for sid in to_close:
                try:
                    await clients[sid]['ws'].close()
                except Exception:
                    pass
                clients.pop(sid, None)
        else:
            await broadcast_to_room(room_code, 'member_left', {'userId': user_id})

        clients.pop(socket_id, None)
        print(f"{user_id} left room {room_code}")
    except Exception as exc:
        print('Error leaving room:', exc)


async def handle_heartbeat(socket_id, payload):
    client = clients.get(socket_id)
    if not client or not client.get('isHost'):
        return

    room_code = client['roomCode']
    await update_room_heartbeat(room_code)


async def handle_search_tracks(socket_id, ws, payload):
    client = clients.get(socket_id)
    if not client:
        return await send_error(ws, 'Not in a room')

    query_text = payload.get('query')
    user_id = client['userId']

    try:
        access_token = await get_valid_access_token(user_id)
        results = await search_tracks(access_token, query_text)
        await send(ws, 'search_results', {'results': results})
    except Exception as exc:
        print('Error searching tracks:', exc)
        await send_error(ws, 'Failed to search tracks')


async def handle_add_to_queue(socket_id, payload):
    client = clients.get(socket_id)
    if not client:
        return

    room_code = client['roomCode']
    user_id = client['userId']
    track = payload.get('track')

    try:
        queue = await add_to_queue(room_code, track, user_id)
        await broadcast_to_room(room_code, 'queue_updated', {'queue': queue})
    except Exception as exc:
        print('Error adding to queue:', exc)

async def handle_remove_from_queue(socket_id, payload):
    client = clients.get(socket_id)
    if not client:
        return

    room_code = client['roomCode']
    is_host = client.get('isHost')
    queue_item_id = payload.get('queueItemId')

    if not is_host:
        return await send_error(client['ws'], 'Only host can remove from queue')

    try:
        queue = await remove_from_queue(room_code, queue_item_id)
        await broadcast_to_room(room_code, 'queue_updated', {'queue': queue})
    except Exception as exc:
        print('Error removing from queue:', exc)


async def handle_playback_control(socket_id, payload):
    client = clients.get(socket_id)
    if not client or not client.get('isHost'):
        return await send_error(client['ws'] if client else None, 'Only host can control playback')

    user_id = client['userId']
    room_code = client['roomCode']
    action = payload.get('action')
    device_id = payload.get('deviceId')
    track_uri = payload.get('trackUri')
    position_ms = payload.get('positionMs')

    try:
        access_token = await get_valid_access_token(user_id)

        if action == 'play':
            await play(access_token, device_id, track_uri, position_ms)
        elif action == 'pause':
            await pause(access_token, device_id)
        elif action == 'next':
            await skip_to_next(access_token, device_id)
        elif action == 'previous':
            await skip_to_previous(access_token, device_id)
        elif action == 'seek':
            await seek(access_token, position_ms, device_id)

        await broadcast_to_room(
            room_code,
            'playback_changed',
            {
                'action': action,
                'deviceId': device_id,
                'trackUri': track_uri,
                'positionMs': position_ms,
            },
        )
    except Exception as exc:
        print('Error controlling playback:', exc)
        await send_error(client['ws'], 'Failed to control playback')


async def handle_sync_playback(socket_id, payload):
    client = clients.get(socket_id)
    if not client or not client.get('isHost'):
        return

    room_code = client['roomCode']
    state = payload.get('state')

    try:
        await update_room_playback_state(room_code, state)
        await broadcast_to_room(room_code, 'playback_state', state, exclude_socket_id=socket_id)
    except Exception as exc:
        print('Error syncing playback:', exc)


async def handle_transfer_device(socket_id, payload):
    client = clients.get(socket_id)
    if not client or not client.get('isHost'):
        return await send_error(client['ws'] if client else None, 'Only host can transfer device')

    user_id = client['userId']
    room_code = client['roomCode']
    device_id = payload.get('deviceId')

    try:
        access_token = await get_valid_access_token(user_id)
        await transfer_playback(access_token, device_id)
        await send(client['ws'], 'device_transferred', {'deviceId': device_id})
        await update_room_playback_state(room_code, {'deviceId': device_id})
    except Exception as exc:
        print('Error transferring device:', exc)
        await send_error(client['ws'], 'Failed to transfer device')

async def handle_request_token(socket_id, ws, payload):
    client = clients.get(socket_id)
    if not client:
        return await send_error(ws, 'Not connected')

    user_id = client['userId']

    try:
        access_token = await get_valid_access_token(user_id)
        await send(ws, 'token_response', {'accessToken': access_token})
    except Exception as exc:
        print('Error getting token:', exc)
        await send_error(ws, 'Failed to get access token')


async def handle_disconnect(socket_id):
    client = clients.get(socket_id)
    if client:
        await handle_leave_room(socket_id)

    clients.pop(socket_id, None)
    print(f"WebSocket disconnected: {socket_id}")


async def send(ws: WebSocket, msg_type: str, payload: dict):
    if ws is None:
        return
    await ws.send_text(json.dumps({'type': msg_type, 'payload': payload}))


async def send_error(ws: WebSocket, message: str):
    await send(ws, 'error', {'message': message})


async def broadcast_to_room(room_code, msg_type, payload, exclude_socket_id=None):
    for socket_id, client in list(clients.items()):
        if client['roomCode'] == room_code and socket_id != exclude_socket_id:
            await send(client['ws'], msg_type, payload)


def generate_socket_id():
    return f"ws_{int(time.time() * 1000)}_{random.random().__str__()[2:11]}"
